File: apps/mobile/lib/core/util/serial.dart
// Frame-serial comparison for item_resolutions serial ranges.
// Dart port of apps/backend/src/lib/serial.ts — keep the two in sync.
//
// Rule: when BOTH values are all-digits (after trim) compare numerically,
// otherwise case-insensitive lexicographic. Honda frame serials in the
// catalogs are all-digit, so the numeric path dominates.

final _allDigits = RegExp(r'^\d+$');

int compareSerial(String a, String b) {
  final ta = a.trim();
  final tb = b.trim();
  if (_allDigits.hasMatch(ta) && _allDigits.hasMatch(tb)) {
    return BigInt.parse(ta).compareTo(BigInt.parse(tb));
  }
  return ta.toLowerCase().compareTo(tb.toLowerCase());
}

/// True when [serial] falls inside [from, to]; a null bound is open on that side.
bool serialInRange(String serial, String? from, String? to) {
  if (from != null && compareSerial(serial, from) < 0) return false;
  if (to != null && compareSerial(serial, to) > 0) return false;
  return true;
}

/// Human label for a resolution's applicability, e.g. "CBS · s/n 1000001–1099999".
/// Null when unrestricted (fits every variant and serial).
String? applicabilityLabel({String? variantName, String? serialFrom, String? serialTo}) {
  final parts = <String>[
    ?variantName,
    if (serialFrom != null || serialTo != null) 's/n ${serialFrom ?? ''}–${serialTo ?? ''}',
  ];
  return parts.isEmpty ? null : parts.join(' · ');
}

File: apps/backend/src/lib/serial.ts
// Frame-serial comparison for item_resolutions serial ranges.
//
// Rule: when BOTH values are all-digits (after trim) compare numerically, otherwise
// case-insensitive lexicographic. Honda frame serials in the catalogs are all-digit,
// so the numeric path dominates; the lexicographic fallback can misorder mixed-width
// alphanumeric serials. Documented in docs/schema.md.

const ALL_DIGITS = /^\d+$/;

export function compareSerial(a: string, b: string): number {
  const ta = a.trim();
  const tb = b.trim();
  if (ALL_DIGITS.test(ta) && ALL_DIGITS.test(tb)) {
    const na = BigInt(ta);
    const nb = BigInt(tb);
    return na < nb ? -1 : na > nb ? 1 : 0;
  }
  const la = ta.toLowerCase();
  const lb = tb.toLowerCase();
  return la < lb ? -1 : la > lb ? 1 : 0;
}

/** True when serial falls inside [from, to]; a null bound is open on that side. */
export function serialInRange(serial: string, from: string | null, to: string | null): boolean {
  if (from !== null && compareSerial(serial, from) < 0) return false;
  if (to !== null && compareSerial(serial, to) > 0) return false;
  return true;
}
